### ISSUE TITLE
Can't get config to work

### ISSUE THREAD

**Issue by PelagicDev** [merdok/homebridge-webos-tv#78]
I'm trying to get the config setup to connect to my TV and I'm clearly missing something. I used [this guide](https://timleland.com/setup-homebridge-to-start-on-bootup/) to setup homebridge, so my config.json is located in `/var/homebridge/` directory instead of `/home/pi/.homebridge`.

Here's my accessory config:
```
"accessories": [
 {
      "accessory": "webostv",
      "name": "My webOS TV",
      "ip": "192.168.1.27",
      "mac": "XXXXXXXXXXX",
      "keyFile": "",
      "volumeControl": "slider",
      "channelControl": false,
      "mediaControl": false,
      "pollingEnabled": true,
      "appSwitch":[
         "com.webos.app.hdmi2",
         "com.webos.app.externalinput.component"
      ]
    }
  ],
```

Here is the log file for the error:
```
Dec 09 20:21:16 raspberrypi systemd[1]: Started Node.js HomeKit Server.
Dec 09 20:21:18 raspberrypi homebridge[4544]: [2018-12-9 20:21:18] Loaded config.json with 1 accessories and 1 platforms.
Dec 09 20:21:18 raspberrypi homebridge[4544]: [2018-12-9 20:21:18] ---
Dec 09 20:21:18 raspberrypi homebridge[4544]: [2018-12-9 20:21:18] Loaded plugin: homebridge-neato
Dec 09 20:21:18 raspberrypi homebridge[4544]: [2018-12-9 20:21:18] Registering platform 'homebridge-neato.NeatoVacuumRobot'
Dec 09 20:21:18 raspberrypi homebridge[4544]: [2018-12-9 20:21:18] ---
Dec 09 20:21:19 raspberrypi homebridge[4544]: [2018-12-9 20:21:19] Loaded plugin: homebridge-webos-tv
Dec 09 20:21:19 raspberrypi homebridge[4544]: [2018-12-9 20:21:19] Registering accessory 'homebridge-webos-tv.webostv'
Dec 09 20:21:19 raspberrypi homebridge[4544]: [2018-12-9 20:21:19] ---
Dec 09 20:21:19 raspberrypi homebridge[4544]: [2018-12-9 20:21:19] Loading 1 platforms...
Dec 09 20:21:19 raspberrypi homebridge[4544]: [2018-12-9 20:21:19] [NeatoVacuumRobot] Initializing NeatoVacuumRobot platform...
Dec 09 20:21:19 raspberrypi homebridge[4544]: Sun, 09 Dec 2018 20:21:19 GMT homebridge-neato Extra Care Navigation: 1
Dec 09 20:21:19 raspberrypi homebridge[4544]: Sun, 09 Dec 2018 20:21:19 GMT homebridge-neato Get all robots
Dec 09 20:21:19 raspberrypi homebridge[4544]: [2018-12-9 20:21:19] Loading 1 accessories...
Dec 09 20:21:19 raspberrypi homebridge[4544]: [2018-12-9 20:21:19] [My webOS TV] Initializing webostv accessory...
Dec 09 20:21:19 raspberrypi homebridge[4544]: path.js:28
Dec 09 20:21:19 raspberrypi homebridge[4544]:     throw new TypeError('Path must be a string. Received ' + inspect(path));
Dec 09 20:21:19 raspberrypi homebridge[4544]:     ^
Dec 09 20:21:19 raspberrypi homebridge[4544]: TypeError: Path must be a string. Received undefined
Dec 09 20:21:19 raspberrypi homebridge[4544]:     at assertPath (path.js:28:11)
Dec 09 20:21:19 raspberrypi homebridge[4544]:     at join (path.js:1236:7)
Dec 09 20:21:19 raspberrypi homebridge[4544]:     at module.exports (/usr/lib/node_modules/homebridge-webos-tv/node_modules/persist-path/index.js:19:22)
Dec 09 20:21:19 raspberrypi homebridge[4544]:     at new LGTV (/usr/lib/node_modules/homebridge-webos-tv/node_modules/lgtv2/index.js:47:16)
Dec 09 20:21:19 raspberrypi homebridge[4544]:     at new webosTvAccessory (/usr/lib/node_modules/homebridge-webos-tv/index.js:55:17)
Dec 09 20:21:19 raspberrypi homebridge[4544]:     at Server._loadAccessories (/usr/lib/node_modules/homebridge/lib/server.js:297:29)
Dec 09 20:21:19 raspberrypi homebridge[4544]:     at Server.run (/usr/lib/node_modules/homebridge/lib/server.js:87:38)
Dec 09 20:21:19 raspberrypi homebridge[4544]:     at module.exports (/usr/lib/node_modules/homebridge/lib/cli.js:45:10)
Dec 09 20:21:19 raspberrypi homebridge[4544]:     at Object.<anonymous> (/usr/lib/node_modules/homebridge/bin/homebridge:17:22)
Dec 09 20:21:19 raspberrypi homebridge[4544]:     at Module._compile (module.js:653:30)
Dec 09 20:21:19 raspberrypi systemd[1]: homebridge.service: Main process exited, code=exited, status=1/FAILURE
Dec 09 20:21:19 raspberrypi systemd[1]: homebridge.service: Unit entered failed state.
Dec 09 20:21:19 raspberrypi systemd[1]: homebridge.service: Failed with result 'exit-code'.
```

**Comment by merdok** [comment]
Remove the `"keyFile": "",` property completely if you do not want to specify it.

**Comment by PelagicDev** [comment]
Yeah, I actually I had tried it that was as well and got the same error. Could that error have anything to do with permissions from the `Homebridge` system user?

**Comment by merdok** [comment]
Could be both, but the error clearly says that the path is empty. Your config.json looks ok (expect the empty keyFile path).

**Comment by PelagicDev** [comment]
I played with it again last night removing the entire `keyfile` line and got the exact same error. Would removing the `persist` folder and doing a re-setup of everything be of any benefit?

**Comment by merdok** [comment]
Then i guess this is not an issue with the plugin but with your setup and/or environment. It might be permissions. You would need to setup everything again from scratch.

**Comment by PelagicDev** [comment]
Yeah, I don't think there's anything wrong with the plugin; there's definitely something wrong with the way I have it setup, just trying to figure out what it is. 
I'll give it another shot tonight.

**Comment by merdok** [comment]
Ok, please let me know what was the issue if you find it.

**Comment by CooperCGN** [comment]
Sorry to reopen this again.

I get exactly the same error:

```
Jan 05 10:42:35 raspberrypi homebridge[9213]: [2019-1-5 10:42:35] [LG OLED E6] Initializing webostv accessory...
Jan 05 10:42:35 raspberrypi homebridge[9213]: TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type undefined
Jan 05 10:42:35 raspberrypi homebridge[9213]:     at assertPath (path.js:39:11)
Jan 05 10:42:35 raspberrypi homebridge[9213]:     at join (path.js:1152:7)
Jan 05 10:42:35 raspberrypi homebridge[9213]:     at module.exports (/usr/local/lib/node_modules/homebridge-webos-tv/node_modules/persist-path/index.js:19:22)
Jan 05 10:42:35 raspberrypi homebridge[9213]:     at new LGTV (/usr/local/lib/node_modules/homebridge-webos-tv/node_modules/lgtv2/index.js:47:16)
Jan 05 10:42:35 raspberrypi homebridge[9213]:     at new webosTvAccessory (/usr/local/lib/node_modules/homebridge-webos-tv/index.js:55:17)
Jan 05 10:42:35 raspberrypi homebridge[9213]:     at Server._loadAccessories (/usr/local/lib/node_modules/homebridge/lib/server.js:297:29)
Jan 05 10:42:35 raspberrypi homebridge[9213]:     at Server.run (/usr/local/lib/node_modules/homebridge/lib/server.js:87:38)
Jan 05 10:42:35 raspberrypi homebridge[9213]:     at module.exports (/usr/local/lib/node_modules/homebridge/lib/cli.js:45:10)
Jan 05 10:42:35 raspberrypi homebridge[9213]:     at Object.<anonymous> (/usr/local/lib/node_modules/homebridge/bin/homebridge:17:22)
Jan 05 10:42:35 raspberrypi homebridge[9213]:     at Module._compile (internal/modules/cjs/loader.js:723:30)
Jan 05 10:42:35 raspberrypi systemd[1]: homebridgelg.service: Main process exited, code=exited, status=1/FAILURE
```

I am running several instances of homebridge and gave your plugin a separate one. So my config.json resides under /var/lib/homebridgeLG and my config.json contains
"keyFile": "/var/lib/homebridgeLG/lgtvKeyFile",.

Whatever I try or even leaving that option blank the above error pops up. I even set read/write/access permission to everyone for the specified folder. Any idea what to do? I really wish I could get it running.

**Comment by CooperCGN** [comment]
Ok I just installed homebridge on a spare Pi I had left. Now it works, I wonder where is the plugin trying to put the file if you leave the option blank. I guess I need to set specific rifts for the homebridge user to that directory.

**Comment by PelagicDev** [comment]
So wait, you just tried a fresh install and it seemed to work this time? Where is your homebridge config file? Is it in /var/homebridge or in the root user dir?

**Comment by merdok** [comment]
@CooperCGN The path for the keyFile is managed by a module that i am using, [lgtv2 ](https://github.com/hobbyquaker/lgtv2)so i actaully do not have control over it:

From the documentation of that module:

> keyFile - path for key storage. Will be suffixed with hostname/ip of TV. default: Linux: ~/.lgtv2/keyfile-, macOS: ~/Library/Preferences/lgtv2/keyfile-

**Comment by CooperCGN** [comment]
My config.json is under /var/lib/homebridge as I’m running homebridge as a systemd service. I assume the module lgtv2 expects rights of a certain user to the folder where it wants to place the keyfile. I guess it needs to be the user homebridge which I created for running the homebridge service

**Comment by CooperCGN** [comment]
I compared the two installations but have no idea why it’s running on one pi but not the other. All access rights to the required folders are the same, maybe I find the time to setup my ho ebridge again later but for now I give up. :(

**Comment by PelagicDev** [comment]
Haha, fair enough. I’ll play around with it and see what I can figure out.

Just curious can you link me to the guide you used to set your working Pi up?

**Comment by CooperCGN** [comment]
I used a newly installed Stretch OS on the Pi downloaded from the raspberry site as an image and then just followed the installation guide here on GitHub on the homebridge site.

**Comment by decepulis** [comment]
Had the same error with a Raspberry Pi 2 (Raspbian), using systemd and the following config,:
 
```js
{
      "accessory": "webostv",
      "name": "Living Room TV",
      "ip": "...",
      "mac": "...",
      "keyFile": "/var/homebridge/webostvKeyFile",
      "prefsDir": "/var/homebridge/webostvPrefsDir",
      ...
}
```

TL;DR... The problem's not in this library. It's not even in lgtv2. It's in a library called persist-path.

Basically, persist-path can't find a good place to store temporary data because it can't find my home directory. To fix this problem, I had to go into my systemd service
```
sudo nano /etc/systemd/system/homebridge.service 
```
And make sure a `User` was defined, like so:
```
[Unit]
Description=Node.js HomeKit Server
After=syslog.target network-online.target

[Service]
Type=simple
User=homebridge
EnvironmentFile=/etc/default/homebridge
ExecStart=/usr/local/bin/homebridge $HOMEBRIDGE_OPTS
Restart=on-failure
RestartSec=10
KillMode=process

[Install]
WantedBy=multi-user.target

```

How did I get there?
<details>
<summary>Here's the unnecessary, long story</summary>

Stuck in some `console.log` in and here's what I found...

The error's being thrown when the Lgtv2 instance is being created.
```js
this.lgtv = new Lgtv2({
            url: this.url,
            timeout: 5000,
            reconnect: 3000,
            keyFile: this.keyFile
        });
```

The `this.keyFile` being passed into the instance, at least in my case, is definitely a string. So, diving deeper... `lgtv2/index.js:47:16`

It looks like the fun happens around here:
```js
    if (typeof config.clientKey === 'undefined') {
        mkdirp(ppath('lgtv2'));
        config.keyFile = (config.keyFile ? config.keyFile : ppath('lgtv2/keyfile-') + config.url.replace(/[a-z]+:\/\/([0-9a-zA-Z-_.]+):[0-9]+/, '$1'));
        try {
            that.clientKey = fs.readFileSync(config.keyFile).toString();
        } catch (err) {}
    } else {
        that.clientKey = config.clientKey;
    }
```

So `config.clientKey` is undefined (expected behavior), so we proceed to `mkdirp(ppath('lgtv2'));`, where we crash. Not too surprising, considering that the stack trace came from `persist-path/index.js:19:22`. So what's going on in there?

Here's the entirety of persist-path.
```js
module.exports = function () {
    var args = Array.prototype.slice.call(arguments);
    var pathArr = [];
    if (process.env.APPDATA) {
        pathArr.push(process.env.APPDATA);
    } else if (process.platform === 'darwin') {
        pathArr.push(path.join(process.env.HOME, 'Library/Preferences'));
    } else {
        pathArr.push(process.env.HOME);
        if (args.length > 0) {
            args[0] = '.' + args[0];
        }
    }
    if (args.length > 0) {
        pathArr = pathArr.concat(args);
    }
    return path.join.apply(this, pathArr);
};
```

 The error is occurring in the `return` statement. Why? because pathArr is defined as `[ undefined, '.lgtv2' ]`. There's our `undefined`! How did we get here? Well, we're actually going to the `else` block at the end of the config because we're not on a `darwin` mac. And I'm not sure what that `process.env.APPDATA` is about. Windows, probably. And of course, because I'm running from systemctl, `process.env.HOME` is `undefined`. Bam. Error.

What do we do now to fix it? Well, the problem's not in this project. It's not even in lgtv2, though it [looks like they're aware](https://github.com/hobbyquaker/lgtv2/issues/27). <sup>If I had found that thread earlier, I could've skipped this whole effort...</sup> It's lower, in persist-path.

And really, all that can be done is run on a mac or in an environment where process.env.APPDATA or process.env.HOME is defined. So how do we define HOME for a systemd service?

We make sure a user is defined.
Which I didn't define in my `/etc/systemd/system/homebridge.service`.

Fixing that fixed the problem.
</details>
